Client/Frontend/Browser/SearchLoader.swift

```diff
@@ -0,0 +1,81 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Shared
+import Storage
+
+private let URLBeforePathRegex = NSRegularExpression(pattern: "^https?://([^/]+/)", options: nil, error: nil)!
+
+// TODO: Swift current requires that classes extending generic classes must also be generic.
+// This is a workaround until that requirement is fixed.
+typealias SearchLoader = _SearchLoader<AnyObject, AnyObject>
+
+/**
+ * Shared data source for the SearchViewController and the URLBar domain completion.
+ * Since both of these use the same SQL query, we can perform the query once and dispatch the results.
+ */
+class _SearchLoader<UnusedA, UnusedB>: Loader<Cursor, SearchViewController> {
+    private let history: History
+    private let urlBar: URLBarView
+
+    init(history: History, urlBar: URLBarView) {
+        self.history = history
+        self.urlBar = urlBar
+        super.init()
+    }
+
+    var query: String = "" {
+        didSet {
+            if query.isEmpty {
+                self.load(Cursor(status: .Success, msg: "Empty query"))
+                return
+            }
+
+            let options = QueryOptions(filter: query, filterType: FilterType.Url, sort: QuerySort.Frecency)
+            self.history.get(options, complete: { (cursor: Cursor) in
+                if cursor.status != .Success {
+                    println("Err: \(cursor.statusMessage)")
+                } else {
+                    self.load(cursor)
+                    self.urlBar.setAutocompleteSuggestion(self.getAutocompleteSuggestion(cursor))
+                }
+            })
+        }
+    }
+
+    private func getAutocompleteSuggestion(cursor: Cursor) -> String? {
```

Comment:
A couple reasons:
 1. The `URLBarView` is just a dumb view that displays its state. I wanted to keep the model separate from the view. I *really* preferred the dataSource/callback-based API I had in the first version of the PR, but that got bastardized thanks to our query sharing optimization.
 1. We can't, thanks to Swift's broken generics. Note how `Loader` requires us to name the listener type, and how `SearchLoader` is a subclass of `Loader<Cursor, SearchViewController>`. We can't simultaneously subclass `Loader<Cursor, URLBarView>`.

Honestly, I'm still not too happy with the way this all turned out, but I haven't been able to come up with much better; the need to reuse the query between completely separate views is a pain. I'll gladly look into any suggestions for improvements in a follow-up.